hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -488,6 +508,78 @@ def invariant_wrapper(*args, **kwargs):
     return accept
 
 
+LOOP_LABEL = cu.calc_label_from_name('RuleStrategy loop iteration')
+
+
+class RuleStrategy(SearchStrategy):
+    def __init__(self, machine):
+        SearchStrategy.__init__(self)
+        self.machine = machine
+        self.rules = list(machine.rules())
+
+        # The order is a bit arbitrary. Primarily we're trying to group rules
+        # that write to the same location together, and to put rules with no
+        # target first as they have less effect on the structure. We order from
+        # fewer to more arguments on grounds that it will plausibly need less
+        # data. This probably won't work especially well and we could be
+        # smarter about it, but it's better than just doing it in definition
+        # order.
+        self.rules.sort(key=lambda rule: (
+            sorted(rule.targets), len(rule.arguments),
+            rule.function.__name__,
+        ))
+
+    def do_draw(self, data):
+        # This strategy is slightly strange in its implementation.
+        # We don't want the interpretation of the rule we draw to change based
+        # on whether other rules satisfy their preconditions or have data in
+        # their bundles. Therefore the index into the rule list needs to stay
+        # stable. BUT we don't want to draw invalid rules. So what we do is we
+        # draw an index. We *could* just loop until it's valid, but if most
+        # rules are invalid then that could result in a very long loop.
+        # So what we do is the following:
+        #
+        #   1. We first draw a rule unconditionally, and check if it's valid.
+        #      If it is, great. Nothing more to do, that's our rule.
+        #   2. If it is invalid, we now calculate the list of valid rules and
+        #      draw from that list (if there are none, that's an error in the
+        #      definition of the machine and we complain to the user about it).
+        #   3. Once we've drawn a valid rule, we write that back to the byte
+        #      stream. As a result, when shrinking runs the shrinker can delete
+        #      the initial failed draw + the draw that lead to us finding an
+        #      index into valid_rules, leaving just the written value of i.
+        #      When this is run, it will look as we got lucky and just happened
+        #      to pick a valid rule.
+        #
+        # Easy, right?
+        n = len(self.rules)
+        i = cu.integer_range(data, 0, n - 1)
+        u, v = data.blocks[-1]
+        block_length = v - u
+        rule = self.rules[i]
+        if not self.is_valid(rule):
+            valid_rules = [
+                j for j, r in enumerate(self.rules) if self.is_valid(r)
+            ]
+            if not valid_rules:
+                raise RuntimeError(
```

Comment:
I've changed this from `InvalidDefinition` to `RuntimeError`, so that the sequence of steps is reported in `find_breaking_runner` rather than immediately re-raising the exception.  In every pre-exisiting case, `InvalidDefinition` is raised at definition time, not runtime.

Alternative options would be to use a different builtin exception, or to special-case `InvalidDefinition` in `find_breaking_runner` so that the breaking steps are still reported.